cmd/swarm/upload.go

```diff
@@ -36,43 +36,93 @@ import (
 	"gopkg.in/urfave/cli.v1"
 )
 
+func cleanupAndDie(file *string, description string) {
+	if file != nil {
+		os.Remove(*file)
+	}
+	utils.Fatalf(description)
+}
+
+func swapStdin(f *os.File) (size int) {
+	b := make([]byte, 1024)
+	path := f.Name()
+	size = 0
+
+	r := io.TeeReader(os.Stdin, f)
+
+	for true {
+		n, err := r.Read(b)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			cleanupAndDie(&path, "STDIN read failed")
+		}
+		size += n
+	}
+	if size == 0 {
+		cleanupAndDie(&path, "STDIN zero length input")
+	}
+	return size
+}
+
 func upload(ctx *cli.Context) {
+
 	args := ctx.Args()
 	var (
 		bzzapi       = strings.TrimRight(ctx.GlobalString(SwarmApiFlag.Name), "/")
 		recursive    = ctx.GlobalBool(SwarmRecursiveUploadFlag.Name)
 		wantManifest = ctx.GlobalBoolT(SwarmWantManifestFlag.Name)
 		defaultPath  = ctx.GlobalString(SwarmUploadDefaultPath.Name)
+		fromStdin    = ctx.GlobalBool(SwarmUpFromStdinFlag.Name)
+		mimeType     = ctx.GlobalString(SwarmUploadMimeType.Name)
 	)
+
+	var client = &client{api: bzzapi}
+	var entry manifestEntry
+	var file string
+	var cleanfile *string
+
 	if len(args) != 1 {
-		utils.Fatalf("Need filename as the first and only argument")
+		if fromStdin {
+			f, err := ioutil.TempFile("", "swarm-stdin")
+			if err != nil {
+				utils.Fatalf("Could not create tempfile")
+			}
+			swapStdin(f)
+			file = f.Name()
+			cleanfile = &file
```

Comment:
Now that you are not limiting the amount of data read from stdin, you can just copy stdin directly to the tempfile here and defer the cleanup, for example:

```go
tmp, err := ioutil.TempFile("", "swarm-stdin")
if err != nil {
	utils.Fatalf("error create tempfile: %s", err)
}
defer os.Remove(tmp.Name())
n, err := io.Copy(tmp, os.Stdin)
if err != nil {
	utils.Fatalf("error copying stdin to tempfile: %s", err)
} else if n == 0 {
	utils.Fatalf("error reading from stdin: zero length")
}
file = tmp.Name()
```

This would remove the need for both the `swapStdin` and `cleanupAndDie` functions.